### DISCUSSION TITLE
Reading json to an array of user-defined objects

### DISCUSSION THREAD

**Discussion by andres-alban** [nlohmann/json#3266] [Q&A]
I want to read a json object to an array of user-defined objects. To improve readability and conciseness of the json file, I would like to write the json as an object with arrays for each entry instead of an array of several objects. For example, suppose I have the following class:
```c++
class BoundedNumber {
public:
    double value{};
    double lower{};
    double upper{};
};
```
and I want to save an array of this class (`std::array<BoundedNumber,3>`) in a json file (ArrayBoundedNumber.json) as follows:
```json
{
	"value":[0,1,2],
	"lower":[-1,0,1],
	"upper":[1,2,3]
}
```
Following the manual, I am overloading `from_json`
```c++
template<size_t L>
void from_json(const json& j, std::array<BoundedNumber, L>& d) {
    for (int i = 0; i < L; ++i) {
        j.at("value").at(i).get_to(d[i].value);
        j.at("upper").at(i).get_to(d[i].upper);
        j.at("lower").at(i).get_to(d[i].lower);
    }
}
```
and reading the file as follows
```c++
int main()
{
    std::ifstream file("ArrayBoundedNumber.json");
    json j;
    file >> j;
    auto d = j.get<std::array<BoundedNumber, 3>>();
    for (int i = 0; i < 3; ++i) {
        std::cout << d[i].lower << " < " << d[i].value << " < " << d[i].upper << "\n";
    }
}
```
This is a MWE of what I had working on my desktop (and it works!) but now I tried to run the same project on my laptop and I am getting a compile error. I am using Visual Studio 2017 and using vcpkg to install nlohmann-json, and I get the compile errors

> Error	C2672	'nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<_Ty>>>::get': no matching overloaded function found	json_test	~\vcpkg\installed\x64-windows\include\nlohmann\detail\conversions\from_json.hpp	272	
> Error	C2893	Failed to specialize function template 'unknown-type nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<_Ty>>>::get(void) noexcept(<expr>) const'	json_test	 ~\vcpkg\installed\x64-windows\include\nlohmann\detail\conversions\from_json.hpp	271	

I tried to remove the template and make it specific to arrays of size 3 but I get the same errors.

The only difference I can tell between my laptop and my desktop is that the version of the nlohmann-json package is 3.10.4 on my laptop and 3.9.1 on my desktop. I would really appreciate if anyone can confirm that the issue lies in the different versions of the package and if there is a solution to have this example work on the latest version.

**Comment by nlohmann** [comment]
The library has a built-in from_json overload for std::array<T, N> (see include/nlohmann/detail/conversions/from_json.hpp, the from_json_array_impl overload taking priority_tag<2>, called from the from_json(BasicJsonType&&, identity_tag<std::array<T,N>>) entry point), which converts a JSON array into std::array<T,N> by calling get<T>() on each element. When you define your own free function from_json(const json&, std::array<BoundedNumber, L>&) to give std::array a different (object-of-arrays) deserialization, it can collide with this built-in overload during overload resolution, producing exactly the 'no matching overloaded function'/'failed to specialize' errors you're seeing - and the details of that collision are sensitive to compiler and library version, matching your observation that it works on 3.9.1 but not 3.10.4. The library doesn't support cleanly re-specializing from_json for container types it already natively handles (std::vector, std::array, std::map, ...); the recommended way is to use a different C++ type for your struct-of-arrays representation (e.g. a small wrapper struct, not std::array directly) and write from_json for that wrapper instead.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
